### ISSUE TITLE
Accessibility improvements 

### ISSUE THREAD

**Issue by simonssspirit** [telerik/kendo-react#209]
This ticket is a summary list of known KendoReact Accessibility Issues. Tracked in one place for simplicity and share-ability with other Kendo teams. Feel free to create a separate ticket if needed. I will mark from red to green the "done" items

:red_circle: We should add an Accessibility theme to the Kendo UI Themes suit. Synced with Ivan Zhekov, this item is here just for completeness and FYI
- contrast
- prominent focus
- always specify a background color

:red_circle: Our components should support an optional property that adds a tabindex attribute to the following HTML elements -All Anchor, AREA, BUTTON, INPUT, OBJECT, SELECT, and TEXTAREA. This is not done for the following components
- Conversational ui
- Dialog
- Grid
- Menu
- PanelBar
- Splitter

:red_circle: Our components should support an optional property that adds Accesskey attribute (aka keyboard shortcut) to the following HTML elements - A, AREA, BUTTON, INPUT, LABEL, LEGEND, and TEXTAREA.
- Almost all components
- [Spec Link](https://www.w3.org/TR/WCAG10-HTML-TECHS/#link-accesskey)

:red_circle: SVG elements should use title or desc or valid ARIA attributes for providing a description. Either aria-labelledby, aria-describedby or aria-label must have a value. Noticed in:
-  Charts
- Gauges
- "enter" element in conversational-ui
- [Spec Link](https://www.access-board.gov/guidelines-and-standards/communications-and-it/about-the-section-508-standards/guide-to-the-section-508-standards/web-based-intranet-and-internet-information-and-applications-1194-22)

:red_circle: IMG elements are required to have a size specified if they have an intrinsic size. The following components are reported to miss height attribute:
- Avatars in Conversational IU
- Buttons with IMG
- Tabstrip with images
- [Spec Link](https://www.w3.org/TR/mobile-accessibility-mapping/)

:red_circle: All IMG elements are required to contain either the ALT or LONGDESC attribute
- Avatar IMGs in Conversational UI
- Tabstrip with images

:red_circle: INPUT element does not use an ALT attribute, TITLE attribute or linked LABEL.
- Conversational UI
- Dropdowns
- Grid Filter subcomponent
- [Spec Link 1](https://www.w3.org/TR/WCAG20-TECHS/H65.html)
- [Spec Link 2](https://www.w3.org/TR/WCAG10-HTML-TECHS/#forms-graphical-buttons)

:red_circle: SELECT element needs label (preferred) or title.
- Dropdowns (ex: e2e/label.html , dropdownlist/select-cases.html)
- Grid - pagination and filter subcomponents
- [Spec Link](https://www.w3.org/TR/WCAG10-HTML-TECHS/#forms-labels)

:red_circle: LINKS / ANCHORS are required to have descriptive texts.
- 4 anchors/links in grid pagination subcomponent
- Advert pics in Conversational UI
- [Spec Link 1](https://www.w3.org/TR/WCAG20-TECHS/H30.html)
- [Spec Link 2](https://www.w3.org/TR/WCAG20-TECHS/H83.html)

:red_circle: All LINKS / ANCHORS not surrounding images are required to not be directly adjacent
- 4 anchors/links in grid pagination subcomponent
- [Spec Link](https://www.w3.org/TR/WCAG10-HTML-TECHS/#group-bypass)
- [Suggestion - maybe group in nav element?](https://www.w3.org/TR/WCAG20-TECHS/H97.html)

:red_circle: LINKS / ANCHORS require ALT attribute
- Close button in a dialog

:red_circle: Data table element must have a caption element
- Calendar
- Multiview Calendar
- Grid
- [Spec Link](https://www.w3.org/TR/WCAG20-TECHS/H39.html)

:red_circle: Use id and headers attribute to associate data cells with header cells in data tables	Complex data table doesn't have headers attributes to TD and TH elements.
- Calendar
- Multiview Calendar
- Grid
- [Spec Link](https://www.w3.org/TR/WCAG20-TECHS/H43)

:red_circle: Use the scope attribute to associate header cells and data cells in data tables	TH elements must specify scope attribute
- Calendar
- Multiview Calendar
- Grid
- [Spec Link](https://www.w3.org/TR/WCAG20-TECHS/H63)

:red_circle: For data tables that have two or more logical levels of row or column headers, use markup to associate data cells and header cells	Table uses markup to associate data and header cells but has no summary.
- Calendar
- Multiview Calendar
- Grid
- [Spec Link](https://www.w3.org/TR/WCAG10-HTML-TECHS/#identifying-table-rows-columns)

:red_circle: For data tables that have two or more logical levels of row or column headers, use markup to associate data cells and header cells.	The table uses markup to associate data and header cells but has no valid ARIA attribute. 
- Calendar
- Multiview Calendar
- Grid
- [Spec Link](https://www.access-board.gov/guidelines-and-standards/communications-and-it/about-the-section-508-standards/guide-to-the-section-508-standards/web-based-intranet-and-internet-information-and-applications-1194-22)

:red_circle: Multiselect renders a list of selected items. We would like to not render the list when it's empty.

:red_circle: As advised by Compliance Sheriff, we ran the rendered HTML of most of our examples against [a W3C Validator](https://validator.w3.org/) and the output contained multiple errors. 
- [Spec Link](https://www.w3.org/TR/WCAG20-TECHS/G134.html)

:red_circle: It is expected that without CSS our components are not pretty, but they should have basic functionality. Here is a list of issues when we run examples without CSS
- Zoom and push animations, popups, tooltips - all play at wrong locations
- Buttons - usable, but not clear if pressed or not
- All rtl demos have issues
- Dropdownlist, Window, Charts, Calendar, Multiview Calendar, Timepicker, Gauges, Switches - broken, not usable
- Dateinput, Datepicker, Daterangepicker, Menu, Panelbar, Splitter, Tabstrip - somewhat usable but buggy
- All dateinputs - gulp often blows with OOM for examples without CSS
- Dialog takes the whole page but works
- Grid - usable but layout is broken
- Input and NumericTextBox work fine just like basic inputs
- Ripple without CSS breaks UI
- The rest are usable enough
- [Spec Link](https://www.w3.org/TR/WCAG10-CSS-TECHS/#style-transform-gracefully)

**Comment by siliconeidolon** [comment]
Hi, are any of these changes likely to be in the September update?

**Comment by RJForgie** [comment]
Any update on when any of these issues will be addressed?

**Comment by sstojanovic-tp** [comment]
Any update on this one? Thanks

**Comment by simonssspirit** [comment]
https://github.com/telerik/kendo-react/issues/814

**Comment by simonssspirit** [comment]
Some of the issues were already addressed in a separate item.

If there is a specific issue in a certain component, please let us know and we will log separate tasks for that case. 

This will allow easy monitoring.

**Comment by melb47** [comment]
Is there any update about adding an accessibility theme?

**Comment by simonssspirit** [comment]
@melb47 

most of these issues are already addressed in different versions.

Please let me know if there is a specific accessibility issue that you have?

**Comment by melb47** [comment]
@simonssspirit I'm looking at the accessibility theme in particular. None of the three provided themes provide the right amount of color contrast on pagination of data grids; Bootstrap and Material are close, however, the inactive arrow elements don't pass.

**Comment by simonssspirit** [comment]
@melb47 Thank you for the clarification.

Regarding the different themes:

1- Default - We are discussing changing the default theme main color in order to achieve an AAA contrast rating. This is still in discussion phase, but we are already trying to address it.

2 - Material and Bootstrap - For these themes, we are following the official material and bootstrap color pallets. That means we will change the colors only if they change otherwise we will not be following their guidance.

**Comment by kdikov82** [comment]
Grid 
- selectAll checkbox needs aria-label to provide meaningful information 
- we could add aria-label to the select checkboxes in each row as well
